### PR TITLE
dynamic client creation rule

### PR DESCRIPTION
**Pull request by Njeriiii**
What does this implement? Explain your changes.
---------------------------------------------------
 This class is used to check for the dynamic creation of clients in the code.
 It extends the LocalInspectionTool class, which is used to create custom code inspections.
 The visitor inspects a for loop and checks for methodcall expressions that create clients using the buildClient or AsyncBuildClient methods.
 If a client creation method is found building a client from the com.azure package, a problem is registered.

Has this been tested?
---------------------------
- [x] Tested

Any relevant logs, screenshots, error output, etc.?
-------------------------------------
Screenshot 2024-07-01 174943

In screenshot 2,
First example is an expression statement. The client is declared at the top before being defined within the loop.
The second example is a declaration statement. The client is declared and defined within the loop
Screenshot 2024-07-01 175026

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/.idea/misc.xml
Comment (Njeriiii) [inline PluginsAndFeatures/azure-toolkit-for-intellij/.idea/misc.xml:1]: Not sure why this is being commited .idea/ directory is set to ignore in the .gitignore and all other .xml files within the directory are ignored too

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+        /**
+         * This method checks for the dynamic creation of clients in the code block of a for loop.
+         * Each assignment statement and declaration statement is checked for the creation of clients.
+         * @param blockChild
+         */
+
+        private void checkClientCreation(PsiStatement blockChild) {
+
+            // This is a check for the expression statement
```
Comment (Njeriiii) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java:121]: will add code examples of expression statements & declaration statements, for now, I've put an explanation in the PR description

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheckTest.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+     *
+     * This tests blocks of code that are a PsiExpressionStatement in the checkClientCreation method.
+     */
+    @Test
+    void testDynamicClientCreationWithExpressionStatement() {
+
+        // assert visitor
+        assertVisitor();
+
+        int NUMBER_OF_INVOCATIONS = 1;
```
Comment (Njeriiii) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheckTest.java]: will rename local variables to camelcase

File: PluginsAndFeatures/azure-toolkit-for-intellij/.idea/misc.xml
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/.idea/misc.xml:1]: figured it out --  there's 2 .idea directories so its the outer-level one that wasn't being ignored

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/README.md [hunk 3]
```diff
@@ -22,4 +22,20 @@ The Java Code Quality Analyzer, is a plugin designed to improve the quality of J
...
 - **Severity: INFO**
 - **Recommendation**: Instead of using the low-level `ServiceBusReceiverAsyncClient`, it's recommended to use the `ServiceBusProcessorClient`. The `ServiceBusProcessorClient` is a higher-level abstraction that simplifies message consumption. It's designed for most common use cases and should be the primary choice for consuming messages. This makes it a more suitable option for most developers and scenarios. 
-Please refer to the [Azure SDK for Java documentation](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/README.md#when-to-use-servicebusprocessorclient) for additional information.
\ No newline at end of file
+Please refer to the [Azure SDK for Java documentation](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/azure-messaging-servicebus/README.md#when-to-use-servicebusprocessorclient) for additional information.
+
+2. #### Use ServiceBusProcessorClient instead of ServiceBusReceiverAsyncClient.
```
Comment (lmolkova) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/README.md]: nit: it seems it's not related to this PR and snuck in from a different one.

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+                // Extract the assignment expression
+                PsiAssignmentExpression assignment = (PsiAssignmentExpression) expression;
+
+                // Extract the right-hand side of the assignment
+                PsiExpression rhs = assignment.getRExpression();
+
+                // Check if the right-hand side is a method call expression
+                if (rhs != null) {
+                    if (isClientCreationMethod((PsiMethodCallExpression) rhs)) {
```
Comment (lmolkova) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java]: super-nit
```suggestion
                if (rhs != null && isClientCreationMethod((PsiMethodCallExpression) rhs)) {
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+        /**
+         * This method checks if the method call expression is a client creation method.
+         * It checks the method name and the type of the qualifier expression.
+         * If the method name is buildClient or AsyncBuildClient and the qualifier expression is of type com.azure,
+         * then it is considered a client creation method.
+         * @param methodCallExpression
+         * @return
+         */
+        public boolean isClientCreationMethod (PsiMethodCallExpression methodCallExpression){
```
Comment (lmolkova) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java]:
```suggestion
        public boolean isClientCreationMethod(PsiMethodCallExpression methodCallExpression){
```

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+            //load json object
+            JSONObject jsonObject = LoadJsonConfigFile.getInstance().getJsonObject();
+
+            //get the async return types to check
+            JSONArray asyncReturnTypes = jsonObject.getJSONObject(RULE_NAME).getJSONArray(METHODS_TO_CHECK_KEY);
+
+            // extract string from json object
+            String antiPatternMessage = jsonObject.getJSONObject(RULE_NAME).getString(ANTI_PATTERN_MESSAGE_KEY);
+
+            return Map.of(METHODS_TO_CHECK_KEY, asyncReturnTypes.toList(), ANTI_PATTERN_MESSAGE_KEY, antiPatternMessage.toString());
```
Comment (lmolkova) [inline PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java]:
```suggestion
            return Map.of(METHODS_TO_CHECK_KEY, asyncReturnTypes.toList(), ANTI_PATTERN_MESSAGE_KEY, antiPatternMessage);
```

**Review by lmolkova — APPROVED** [review]
Left a few minor comments, looks great otherwise! :)

File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java (hunk 6 above)
Comment (Njeriiii) [reply PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/DynamicClientCreationCheck.java]: this section is refactored now:)

**Pull request merged**